### PR TITLE
Overhaul the `-l` option parser (for linking to native libs)

### PR DESCRIPTION
**Pull request by Zalathar**
The current parser for `-l` options has accumulated over time, making it hard to follow. This PR tries to clean it up in several ways.

Key changes:
- This code now gets its own submodule, to slightly reduce clutter in `rustc_session::config`.
- Cleaner division between iterating over multiple `-l` options, and processing each individual one.
- Separate “split” step that breaks up the value string into `[KIND[:MODIFIERS]=]NAME[:NEW_NAME]`, but leaves parsing/validating those parts to later steps.
  - This step also gets its own (disposable) unit test, to make sure it works as expected.
- A context struct reduces the burden of parameter passing, and makes it easier to write error messages that adapt to nightly/stable compilers.
- Fewer calls to `nightly_options` helper functions, because at this point we can get the same information from `UnstableOptions` and `UnstableFeatures` (which are downstream of earlier calls to those helper functions).

There should be no overall change in compiler behaviour.

### CODE REVIEW

**Comment by petrochenkov** [comment]
@Nadrieril 
It was a reminder for myself to take a look, which I indeed took later (there are several hours between assignment and unassignment).

**Comment by Nadrieril** [comment]
r? compiler

**Review by jieyouxu — APPROVED** [review]
Thanks, AFAICT this looks good.

**Comment by jieyouxu** [comment]
@Zalathar one question: do we want any dedicated UI tests for the failure cases? But otherwise r=me.

**Comment by Zalathar** [comment]
(Rebased; no changes.)

Still looking into the test situation; it does seem a bit worrying.

**Comment by jieyouxu** [comment]
Yeah. AFAICT this has very sparse coverage (esp. for the more complex cases). It would be very cool if we can figure out some way to test this (like make the intermediate parsing outcome unit testable).

**Comment by Zalathar** [comment]
Decided that I don't want to fall too far down the rabbit-hole of adding tests, so I added just a few that I noticed were conspicuously missing [(diff)](https://github.com/rust-lang/rust/compare/5e8fed2747fa7541de80cafa1b7f00bdf5e97f90..78edefea9d8e4820d7ce7c1d846f89e60fe3b81c).

@rustbot ready

**Comment by Zalathar** [comment]
> It would be very cool if we can figure out some way to test this (like making the intermediate parsing outcome unit testable before it gets fed to library searching).

I do have some unit tests for one of the important intermediate steps (`split_native_lib_value`). Adding unit tests for `NativeLib` output should be viable, though I'll leave that for future work.
